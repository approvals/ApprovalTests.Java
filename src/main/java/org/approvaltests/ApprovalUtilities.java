package org.approvaltests;

import java.io.ByteArrayOutputStream;
import java.io.PrintStream;

public class ApprovalUtilities
{
  public ByteArrayOutputStream writeSystemOutToStringBuffer()
  {
    ByteArrayOutputStream text = new ByteArrayOutputStream();
    System.setOut(new PrintStream(text));
    return text;
  }
}
