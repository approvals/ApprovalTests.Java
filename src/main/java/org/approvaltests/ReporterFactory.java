package org.approvaltests;

import java.lang.annotation.Annotation;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import org.approvaltests.core.ApprovalFailureReporter;
import org.approvaltests.reporters.DiffReporter;
import org.approvaltests.reporters.FileLauncherReporter;
import org.approvaltests.reporters.ImageReporter;
import org.approvaltests.reporters.MultiReporter;
import org.approvaltests.reporters.QuietReporter;
import org.approvaltests.reporters.UseReporter;

import com.spun.util.ClassUtils;

public class ReporterFactory
{
  private static HashMap<String, Class<? extends ApprovalFailureReporter>> reporters = new HashMap<String, Class<? extends ApprovalFailureReporter>>();
  public static class FileTypes
  {
    public static final String  Text    = "txt";
    public static final String  Html    = "html";
    public static final String  Excel   = "csv";
    public static final String  File    = "file";
    public static final String  Image   = "png";
    private static final String Default = "default";
  }
  static
  {
    setupReporters();
  }
  public static ApprovalFailureReporter get(String string)
  {
    ApprovalFailureReporter returned = getFromAnnotation();
    returned = tryFor(returned, reporters.get(string));
    returned = tryFor(returned, reporters.get(FileTypes.Default));
    return returned;
  }
  public static ApprovalFailureReporter getFromAnnotation()
  {
    UseReporter reporter = getAnnotationFromStackTrace(UseReporter.class);
    return reporter == null ? null : getReporter(reporter);
  }
  private static ApprovalFailureReporter getReporter(UseReporter reporter)
  {
    Class<? extends ApprovalFailureReporter>[] classes = reporter.value();
    List<ApprovalFailureReporter> reporters = new ArrayList<ApprovalFailureReporter>();
    for (Class<? extends ApprovalFailureReporter> clazz : classes)
    {
      ApprovalFailureReporter instance = ClassUtils.create(clazz);
      reporters.add(instance);
    }
    return reporters.size() == 1 ? reporters.get(0) : new MultiReporter(reporters);
  }
  private static <T extends Annotation> T getAnnotationFromStackTrace(Class<T> annotationClass)
  {
    StackTraceElement[] trace = Thread.currentThread().getStackTrace();
    for (StackTraceElement stack : trace)
    {
      Method method = null;
      Class<?> clazz = null;
      try
      {
        String methodName = stack.getMethodName();
        clazz = Class.forName(stack.getClassName());
        method = clazz.getMethod(methodName, null);
      }
      catch (Exception e)
      {
        //ignore
      }
      T annotation = null;
      if (method != null)
      {
        annotation = method.getAnnotation(annotationClass);
      }
      if (annotation != null) { return annotation; }
      annotation = clazz.getAnnotation(annotationClass);
      if (annotation != null) { return annotation; }
    }
    return null;
  }
  private static ApprovalFailureReporter tryFor(ApprovalFailureReporter returned,
      Class<? extends ApprovalFailureReporter> trying)
  {
    if (returned == null && trying != null) { return ClassUtils.create(trying); }
    return returned;
  }
  private static void setupReporters()
  {
    reporters.put(FileTypes.Text, DiffReporter.class);
    reporters.put(FileTypes.Html, DiffReporter.class);
    reporters.put(FileTypes.Excel, FileLauncherReporter.class);
    reporters.put(FileTypes.File, FileLauncherReporter.class);
    reporters.put(FileTypes.Image, ImageReporter.class);
    reporters.put(FileTypes.Default, QuietReporter.class);
  }
  public static void clearAllReportersExceptDefault()
  {
    Class all = reporters.get(FileTypes.Default);
    reporters.clear();
    reporters.put(FileTypes.Default, all);
  }
}
