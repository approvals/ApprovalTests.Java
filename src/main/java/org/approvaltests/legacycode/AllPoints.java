package org.approvaltests.legacycode;

import java.awt.Point;
import java.util.ArrayList;
import java.util.List;

public class AllPoints
{
  public static class MyPoint extends Point
  {
    public MyPoint(int x, int y)
    {
      super(x, y);
    }
    @Override
    public String toString()
    {
      return String.format("[%s,%s]", x, y);
    }
  }
  public static Point[] get(int xStart, int yStart, int xEnd, int yEnd)
  {
    int xLength = xEnd - xStart + 1;
    int yLength = yEnd - yStart + 1;
    List<Point> points = new ArrayList<Point>(xLength * yLength);
    for (int x = xStart; x <= xEnd; x++)
    {
      for (int y = yStart; y <= yEnd; y++)
      {
        points.add(new MyPoint(x, y));
      }
    }
    return points.toArray(new Point[points.size()]);
  }
}
