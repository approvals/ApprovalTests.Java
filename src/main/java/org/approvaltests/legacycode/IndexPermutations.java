package org.approvaltests.legacycode;

import java.util.Iterator;

public class IndexPermutations implements Iterable<Integer[]>, Iterator<Integer[]>
{
  private final Integer[] sizes;
  private Integer[]       index;
  private boolean         finished = false;
  public IndexPermutations(Integer[] sizes)
  {
    this.sizes = sizes;
    this.index = new Integer[sizes.length];
    for (int i = 0; i < sizes.length; i++)
    {
      index[i] = 0;
    }
  }
  @Override
  public Iterator<Integer[]> iterator()
  {
    return this;
  }
  @Override
  public boolean hasNext()
  {
    return !finished;
  }
  @Override
  public Integer[] next()
  {
    Integer[] r = index.clone();
    incermentIndex(0);
    return r;
  }
  private void incermentIndex(int index)
  {
    this.index[index]++;
    if (this.index[index].equals(sizes[index]))
    {
      if (index == sizes.length - 1)
      {
        finished = true;
        return;
      }
      this.index[index] = 0;
      incermentIndex(index + 1);
    }
  }
  @Override
  public void remove()
  {
    throw new RuntimeException("remove not implemented for IndexPermutations");
  }
}
