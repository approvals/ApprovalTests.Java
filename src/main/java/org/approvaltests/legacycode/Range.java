package org.approvaltests.legacycode;

import java.util.ArrayList;

public class Range {

	public static Integer[] get(int start, int stop) {
		ArrayList<Integer> a = new ArrayList<Integer>();
		for (int i = start; i <= stop; i++) {
		a.add(i);
		}
		return (Integer[]) a.toArray(new Integer[a.size()]);
	}

}
