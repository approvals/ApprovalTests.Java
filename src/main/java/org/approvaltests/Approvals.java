package org.approvaltests;

import java.awt.Component;
import java.awt.Image;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.FileFilter;
import java.io.FilenameFilter;
import java.sql.ResultSet;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;

import org.apache.hadoop.fs.Path;
import org.approvaltests.ReporterFactory.FileTypes;
import org.approvaltests.approvers.FileApprover;
import org.approvaltests.core.ApprovalFailureOverrider;
import org.approvaltests.core.ApprovalFailureReporter;
import org.approvaltests.core.ApprovalWriter;
import org.approvaltests.namer.ApprovalNamer;
import org.approvaltests.namer.JUnitStackTraceNamer;
import org.approvaltests.reporters.ExecutableQueryFailure;
import org.approvaltests.writers.ApprovalBinaryFileWriter;
import org.approvaltests.writers.ApprovalTextWriter;
import org.approvaltests.writers.ApprovalXmlWriter;
import org.approvaltests.writers.ComponentApprovalWriter;
import org.approvaltests.writers.DirectoryToDirectoryWriter;
import org.approvaltests.writers.FileApprovalWriter;
import org.approvaltests.writers.ImageApprovalWriter;
import org.approvaltests.writers.ResultSetApprovalWriter;
import org.jrack.RackResponse;
import org.jrack.RackResponseUtils;
import org.lambda.functions.Function1;
import org.lambda.functions.implementations.F1;
import org.lambda.query.Query;

import com.spun.util.ArrayUtils;
import com.spun.util.ObjectUtils;
import com.spun.util.StringUtils;
import com.spun.util.images.ImageWriter;
import com.spun.util.persistence.ExecutableQuery;
import com.spun.util.persistence.SqlLoader;

public class Approvals
{
  /**
   * @deprecated Use verify()
   */
  public static void approve(String response) throws Exception
  {
    verify(response);
  }
  public static void verify(String response) throws Exception
  {
    verify(new ApprovalTextWriter(response, "txt"), FileTypes.Text);
  }
  public static void verify(Object o) throws Exception
  {
    verify("" + o);
  }
  /**
   * @deprecated Use verifyAll()
   */
  public static <T> void approve(String label, T[] array) throws Exception
  {
    verifyAll(label, array);
  }
  public static <T> void verifyAll(String label, T[] array) throws Exception
  {
    verify(new ApprovalTextWriter(StringUtils.toString(label, array), "txt"), FileTypes.Text);
  }
  /**
   * @deprecated Use verifyAll()
   */
  public static <T> void approve(String header, String label, T[] array) throws Exception
  {
    verifyAll(header, label, array);
  }
  public static <T> void verifyAll(String header, String label, T[] array) throws Exception
  {
    verify(new ApprovalTextWriter(formatHeader(header) + StringUtils.toString(label, array), "txt"),
        FileTypes.Text);
  }
  /**
   * @deprecated Use verifyAll()
   */
  public static <T> void approve(T[] values, Function1<T, String> f1)
  {
    verifyAll(values, f1);
  }
  public static <T> void verifyAll(T[] values, Function1<T, String> f1)
  {
    String text = ArrayUtils.toString(values, f1);
    verify(new ApprovalTextWriter(text, "txt"), FileTypes.Text);
  }
  /**
   * @deprecated Use verifyAll()
   */
  public static <T> void approve(String header, T[] values, Function1<T, String> f1)
  {
    verifyAll(header, values, f1);
  }
  public static <T> void verifyAll(String header, T[] values, Function1<T, String> f1)
  {
    verifyAll(header, Arrays.asList(values), f1);
  }
  public static <T> void verifyAll(String header, Iterable<T> array, Function1<T, String> f1)
  {
    String text = formatHeader(header) + ArrayUtils.toString(array, f1);
    verify(new ApprovalTextWriter(text, "txt"), FileTypes.Text);
  }
  private static String formatHeader(String header)
  {
    return StringUtils.isEmpty(header) ? "" : header + "\r\n\r\n\r\n";
  }
  /**
   * @deprecated Use verifyAll()
   */
  public static <T> void approve(String label, Iterable<T> array) throws Exception
  {
    verifyAll(label, array);
  }
  public static <T> void verifyAll(String label, Iterable<T> array) throws Exception
  {
    verify(new ApprovalTextWriter(StringUtils.toString(label, array), "txt"), FileTypes.Text);
  }
  /**
   * @deprecated Use verifyAll()
   */
  public static <T> void approve(String header, String label, Iterable<T> array) throws Exception
  {
    verifyAll(header, label, array);
  }
  public static <T> void verifyAll(String header, String label, Iterable<T> array) throws Exception
  {
    verify(new ApprovalTextWriter(formatHeader(header) + StringUtils.toString(label, array), "txt"),
        FileTypes.Text);
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(Component c) throws Exception
  {
    verify(c);
  }
  public static void verify(Component c) throws Exception
  {
    verify(new ComponentApprovalWriter(c), FileTypes.Image);
  }
  /**
   * @deprecated Use verifyHtml()
   */
  public static void approveHtml(String response) throws Exception
  {
    verifyHtml(response);
  }
  public static void verifyHtml(String response) throws Exception
  {
    verify(new ApprovalTextWriter(response, "html"), FileTypes.Html);
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(File generateFile)
  {
    verify(generateFile);
  }
  public static void verify(File generateFile)
  {
    verify(new FileApprovalWriter(generateFile), FileTypes.File);
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(Image image)
  {
    verify(image);
  }
  public static void verify(Image image)
  {
    approve(ImageWriter.toBufferedImage(image), createApprovalNamer());
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(BufferedImage bufferedImage)
  {
    verify(bufferedImage);
  }
  public static void verify(BufferedImage bufferedImage)
  {
    verify(new ImageApprovalWriter(bufferedImage), FileTypes.Image);
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(ApprovalWriter writter, ApprovalNamer namer, ApprovalFailureReporter reporter)
  {
    verify(writter, namer, reporter);
  }
  public static void verify(ApprovalWriter writter, ApprovalNamer namer, ApprovalFailureReporter reporter)
  {
    verify(new FileApprover(writter, namer), reporter);
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(ApprovalWriter writter, String fileType)
  {
    verify(writter, fileType);
  }
  public static void verify(ApprovalWriter writter, String fileType)
  {
    verify(writter, createApprovalNamer(), ReporterFactory.get(fileType));
  }
  /**
   * @deprecated Use verifyXml()
   */
  public static void approveXml(String xml) throws Exception
  {
    verifyXml(xml);
  }
  public static void verifyXml(String xml) throws Exception
  {
    verify(new ApprovalXmlWriter(xml), FileTypes.Text);
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(FileApprover approver, ApprovalFailureReporter reporter)
  {
    verify(approver, reporter);
  }
  public static void verify(FileApprover approver, ApprovalFailureReporter reporter)
  {
    try
    {
      if (!approver.approve())
      {
        boolean passed = false;
        if (reporter instanceof ApprovalFailureOverrider)
        {
          passed = approver.askToChangeReceivedToApproved((ApprovalFailureOverrider) reporter);
        }
        if (!passed)
        {
          approver.reportFailure(reporter);
          approver.fail();
        }
        else
        {
          approver.cleanUpAfterSuccess(reporter);
        }
      }
      else
      {
        approver.cleanUpAfterSuccess(reporter);
      }
    }
    catch (Exception e)
    {
      throw ObjectUtils.throwAsError(e);
    }
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(ExecutableQuery query) throws Exception
  {
    verify(query);
  }
  public static void verify(ExecutableQuery query) throws Exception
  {
    verify(new ApprovalTextWriter(query.getQuery(), "txt"), new JUnitStackTraceNamer(),
        new ExecutableQueryFailure(query));
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(Map map) throws Exception
  {
    verify(map);
  }
  public static void verify(Map map) throws Exception
  {
    verify(new ApprovalTextWriter(StringUtils.toString(map), "txt"), FileTypes.Text);
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(RackResponse response) throws Exception
  {
    verify(response);
  }
  public static void verify(RackResponse response) throws Exception
  {
    if (isImage(response))
    {
      String fileType = "png";
      verify(new ApprovalBinaryFileWriter(response.getResponse(), fileType), fileType);
    }
    else
    {
      verifyHtml(response.getResponse().toString());
    }
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(ResultSet rs) throws Exception
  {
    verify(rs);
  }
  public static void verify(ResultSet rs) throws Exception
  {
    verify(new ResultSetApprovalWriter(rs), "csv");
  }
  /**
   * @deprecated Use verify()
   */
  public static void approve(SqlLoader loader) throws Exception
  {
    verify(loader);
  }
  public static void verify(SqlLoader loader) throws Exception
  {
    verify(new SqlLoader.ExecutableWrapper(loader));
  }
  private static boolean isImage(RackResponse response)
  {
    String type = response.getHeaders().get(RackResponseUtils.CONTENT_TYPE);
    return RackResponseUtils.CONTENT_TYPE_IMAGE.equals(type);
  }
  public static ApprovalNamer createApprovalNamer()
  {
    return new JUnitStackTraceNamer();
  }
  private static void approve(BufferedImage bufferedImage, ApprovalNamer namer)
  {
    verify(new ImageApprovalWriter(bufferedImage), FileTypes.Image);
  }
  public static void verifyEachFileInDirectory(File directory)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles());
  }
  public static void verifyEachFileInDirectory(File directory, FileFilter filter)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles(filter));
  }
  public static void verifyEachFileInDirectory(File directory, FilenameFilter filter)
  {
    verifyEachFileAgainstMasterDirectory(directory.listFiles(filter));
  }
  private static void verifyEachFileAgainstMasterDirectory(File[] files) throws Error
  {
    ApprovalNamer namer = createApprovalNamer();
    String dirName = namer.getSourceFilePath() + Path.SEPARATOR + namer.getApprovalName() + ".Files";
    File approvedDirectory = new File(dirName);
    List<File> mismatched = new ArrayList<File>();
    for (File f : files)
    {
      if (!f.isDirectory())
      {
        try
        {
          verify(new DirectoryToDirectoryWriter(f, approvedDirectory), FileTypes.File);
        }
        catch (Throwable e)
        {
          mismatched.add(f);
        }
      }
    }
    if (!mismatched.isEmpty())
    {
      String message = "The Following Files did not match up: " + getFileNameList(mismatched);
      throw new Error(message);
    }
  }
  private static String getFileNameList(List<File> mismatched)
  {
    return Query.select(mismatched, new F1<File, String>(mismatched.get(1))
    {
      {
        ret(a.getName());
      }
    }).toString();
  }
}