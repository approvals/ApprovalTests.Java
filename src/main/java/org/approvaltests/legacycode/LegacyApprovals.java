package org.approvaltests.legacycode;

import java.lang.reflect.Method;
import java.util.Arrays;

import org.approvaltests.Approvals;
import org.approvaltests.ReporterFactory;
import org.approvaltests.ReporterFactory.FileTypes;
import org.approvaltests.namer.JUnitStackTraceNamer;
import org.approvaltests.writers.ApprovalTextWriter;

import com.spun.util.MethodExecutionPath.Parameters;

public class LegacyApprovals
{
  public static void LockDown(Object call, String method, Object[]... parametersVariations) throws Exception
  {
    StringBuffer sb = new StringBuffer();
    IndexPermutations perms = new IndexPermutations(getSizes(parametersVariations));
    Method m = null;
    for (Integer[] indexs : perms)
    {
      Object p[] = getParameters(parametersVariations, indexs);
      if (m == null)
      {
        m = new Parameters(p).getBestFitMethod(call.getClass(), method);
      }
      
      Object out;
      try
      {
        out = m.invoke(call, p);
      }
      catch (Throwable t)
      {
        out = t;
      }
      sb.append(String.format("%s = %s \r\n", Arrays.toString(p), out));
    }
    Approvals.verify(new ApprovalTextWriter(sb.toString(), "txt"), new JUnitStackTraceNamer(), ReporterFactory
        .get(FileTypes.Text));
  }
  private static Object[] getParameters(Object[][] parametersVariations, Integer[] index)
  {
    Object[] parameters = new Object[index.length];
    for (int i = 0; i < index.length; i++)
    {
      parameters[i] = parametersVariations[i][index[i]];
    }
    return parameters;
  }
  public static Integer[] getSizes(Object[]... parametersVariations)
  {
    Integer[] sizes = new Integer[parametersVariations.length];
    for (int i = 0; i < sizes.length; i++)
    {
      sizes[i] = parametersVariations[i].length;
    }
    return sizes;
  }
}
