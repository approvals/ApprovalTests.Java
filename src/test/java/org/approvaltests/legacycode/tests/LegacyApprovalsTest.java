package org.approvaltests.legacycode.tests;

import java.awt.Point;

import junit.framework.TestCase;

import org.approvaltests.Approvals;
import org.approvaltests.legacycode.AllPoints;
import org.approvaltests.legacycode.LegacyApprovals;

public class LegacyApprovalsTest extends TestCase
{
  public void testLockDown() throws Exception
  {
    LegacyApprovals.LockDown(this, "processCall", new Integer[]{1, 2, 3, 4, 5}, new String[]{"a", "b", "c", "d"});
  }
  public Object processCall(Integer i, String s)
  {
    return String.format("[%s, %s]", i, s);
  }
  public void testSizes() throws Exception
  {
    Approvals.verifyAll("size", LegacyApprovals.getSizes(new String[9], new String[3], new String[5]));
  }
  public void testLockDownWith3Things() throws Exception
  {
    LegacyApprovals.LockDown(this, "processCall", AllPoints.get(0, 0, 1, 1), new String[]{"a", "b", "c", "d"},
        new Integer[]{1, 2, 3, 4, 5});
  }
  public Object processCall(Point p, String s, Integer i)
  {
    return String.format("[%s, %s, %s]", p, s, i);
  }
}
