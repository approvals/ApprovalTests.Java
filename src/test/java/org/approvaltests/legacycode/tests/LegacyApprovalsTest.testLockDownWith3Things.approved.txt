[[0,0], a, 1] = [[0,0], a, 1] 
[[0,1], a, 1] = [[0,1], a, 1] 
[[1,0], a, 1] = [[1,0], a, 1] 
[[1,1], a, 1] = [[1,1], a, 1] 
[[0,0], b, 1] = [[0,0], b, 1] 
[[0,1], b, 1] = [[0,1], b, 1] 
[[1,0], b, 1] = [[1,0], b, 1] 
[[1,1], b, 1] = [[1,1], b, 1] 
[[0,0], c, 1] = [[0,0], c, 1] 
[[0,1], c, 1] = [[0,1], c, 1] 
[[1,0], c, 1] = [[1,0], c, 1] 
[[1,1], c, 1] = [[1,1], c, 1] 
[[0,0], d, 1] = [[0,0], d, 1] 
[[0,1], d, 1] = [[0,1], d, 1] 
[[1,0], d, 1] = [[1,0], d, 1] 
[[1,1], d, 1] = [[1,1], d, 1] 
[[0,0], a, 2] = [[0,0], a, 2] 
[[0,1], a, 2] = [[0,1], a, 2] 
[[1,0], a, 2] = [[1,0], a, 2] 
[[1,1], a, 2] = [[1,1], a, 2] 
[[0,0], b, 2] = [[0,0], b, 2] 
[[0,1], b, 2] = [[0,1], b, 2] 
[[1,0], b, 2] = [[1,0], b, 2] 
[[1,1], b, 2] = [[1,1], b, 2] 
[[0,0], c, 2] = [[0,0], c, 2] 
[[0,1], c, 2] = [[0,1], c, 2] 
[[1,0], c, 2] = [[1,0], c, 2] 
[[1,1], c, 2] = [[1,1], c, 2] 
[[0,0], d, 2] = [[0,0], d, 2] 
[[0,1], d, 2] = [[0,1], d, 2] 
[[1,0], d, 2] = [[1,0], d, 2] 
[[1,1], d, 2] = [[1,1], d, 2] 
[[0,0], a, 3] = [[0,0], a, 3] 
[[0,1], a, 3] = [[0,1], a, 3] 
[[1,0], a, 3] = [[1,0], a, 3] 
[[1,1], a, 3] = [[1,1], a, 3] 
[[0,0], b, 3] = [[0,0], b, 3] 
[[0,1], b, 3] = [[0,1], b, 3] 
[[1,0], b, 3] = [[1,0], b, 3] 
[[1,1], b, 3] = [[1,1], b, 3] 
[[0,0], c, 3] = [[0,0], c, 3] 
[[0,1], c, 3] = [[0,1], c, 3] 
[[1,0], c, 3] = [[1,0], c, 3] 
[[1,1], c, 3] = [[1,1], c, 3] 
[[0,0], d, 3] = [[0,0], d, 3] 
[[0,1], d, 3] = [[0,1], d, 3] 
[[1,0], d, 3] = [[1,0], d, 3] 
[[1,1], d, 3] = [[1,1], d, 3] 
[[0,0], a, 4] = [[0,0], a, 4] 
[[0,1], a, 4] = [[0,1], a, 4] 
[[1,0], a, 4] = [[1,0], a, 4] 
[[1,1], a, 4] = [[1,1], a, 4] 
[[0,0], b, 4] = [[0,0], b, 4] 
[[0,1], b, 4] = [[0,1], b, 4] 
[[1,0], b, 4] = [[1,0], b, 4] 
[[1,1], b, 4] = [[1,1], b, 4] 
[[0,0], c, 4] = [[0,0], c, 4] 
[[0,1], c, 4] = [[0,1], c, 4] 
[[1,0], c, 4] = [[1,0], c, 4] 
[[1,1], c, 4] = [[1,1], c, 4] 
[[0,0], d, 4] = [[0,0], d, 4] 
[[0,1], d, 4] = [[0,1], d, 4] 
[[1,0], d, 4] = [[1,0], d, 4] 
[[1,1], d, 4] = [[1,1], d, 4] 
[[0,0], a, 5] = [[0,0], a, 5] 
[[0,1], a, 5] = [[0,1], a, 5] 
[[1,0], a, 5] = [[1,0], a, 5] 
[[1,1], a, 5] = [[1,1], a, 5] 
[[0,0], b, 5] = [[0,0], b, 5] 
[[0,1], b, 5] = [[0,1], b, 5] 
[[1,0], b, 5] = [[1,0], b, 5] 
[[1,1], b, 5] = [[1,1], b, 5] 
[[0,0], c, 5] = [[0,0], c, 5] 
[[0,1], c, 5] = [[0,1], c, 5] 
[[1,0], c, 5] = [[1,0], c, 5] 
[[1,1], c, 5] = [[1,1], c, 5] 
[[0,0], d, 5] = [[0,0], d, 5] 
[[0,1], d, 5] = [[0,1], d, 5] 
[[1,0], d, 5] = [[1,0], d, 5] 
[[1,1], d, 5] = [[1,1], d, 5] 
